tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc

```diff
@@ -0,0 +1,674 @@
+          /*stream=*/cu_stream);
+      if (err != 0) {
+        return errors::Internal(
+            "SparseFillEmptyRows: Could not launch "
+            "gpuprim::DeviceScan::InclusiveSum to scan elements_per_row, "
+            "temp_storage_bytes:",
+            temp_storage_bytes, ", status: ", cudaGetErrorString(err));
+      }
+    }
+
+    Tensor empty_row_indicator_t;
+    bool* empty_row_indicator;
+    if (context->output_required(kEmptyRowIndicatorOutput)) {
+      Tensor* empty_row_indicator_t_ptr = nullptr;
+      TF_RETURN_IF_ERROR(context->allocate_output(kEmptyRowIndicatorOutput,
+                                                  TensorShape({dense_rows}),
+                                                  &empty_row_indicator_t_ptr));
+      empty_row_indicator = empty_row_indicator_t_ptr->vec<bool>().data();
+    } else {
+      TF_RETURN_IF_ERROR(context->allocate_temp(
+          DT_BOOL, TensorShape({dense_rows}), &empty_row_indicator_t));
+      empty_row_indicator = empty_row_indicator_t.vec<bool>().data();
+    }
+
+    TF_RETURN_IF_ERROR(wrap_kernel_call(ComputeEmptyRowIndicatorKernel<Tindex>,
+                                        device, dense_rows, elements_per_row,
+                                        empty_row_indicator));
+
+    // For each row, the number of empty rows up to and including that row.
+    Tensor num_empty_rows_through_t;
+    TF_RETURN_IF_ERROR(context->allocate_temp(
+        index_type, TensorShape({dense_rows}), &num_empty_rows_through_t));
+    auto num_empty_rows_through = num_empty_rows_through_t.flat<Tindex>();
+
+    err = gpuprim::DeviceScan::InclusiveSum(
+        /*d_temp_storage=*/nullptr,
+        /*temp_storage_bytes=*/temp_storage_bytes,
+        /*d_in=*/empty_row_indicator,
+        /*d_out=*/num_empty_rows_through.data(),
+        /*num_items=*/dense_rows,
+        /*stream=*/cu_stream);
+    if (err != 0) {
+      return errors::Internal(
+          "SparseFillEmptyRows: Could not launch "
+          "gpuprim::DeviceScan::ExclusiveSum to calculate temp_storage_bytes, "
+          "status: ",
+          cudaGetErrorString(err));
+    }
+    {
+      Tensor temp_storage2;
+      TF_RETURN_IF_ERROR(context->allocate_temp(
+          DT_INT8, TensorShape({static_cast<int64>(temp_storage_bytes)}),
+          &temp_storage2));
+      err = gpuprim::DeviceScan::InclusiveSum(
+          /*d_temp_storage=*/temp_storage2.flat<int8>().data(),
+          /*temp_storage_bytes=*/temp_storage_bytes,
+          /*d_in=*/empty_row_indicator,
+          /*d_out=*/num_empty_rows_through.data(),
+          /*num_items=*/dense_rows,
+          /*stream=*/cu_stream);
+      if (err != 0) {
+        return errors::Internal(
+            "SparseFillEmptyRows: Could not launch "
+            "gpuprim::DeviceScan::ExclusiveSum to scan empty_row_indicator, "
+            "temp_storage_bytes:",
+            temp_storage_bytes, ", status: ", cudaGetErrorString(err));
+      }
+    }
```

Comment:
This routine (first get temp bytes, allocate temp byte and do actual work) can be a helper function.